### PR TITLE
Support padding on every dimension of a rank > 4 tensor in ttnn.pad

### LINKED ISSUE #38144: ttnn::pad only supports padding on the lowest 3 dimensions for tensors with rank > 4
ttnn::pad only supports padding on the lowest 3 dimensions for tensors with rank > 4. Ideally it should support all dimensions. 
Following error occurs:
Always | TT_FATAL: ttnn::pad only supports padding on the lowest 3 dimensions for tensors with rank > 4

Python script to reproduce the issue:
```
import ttnn
import torch

device = ttnn.open_device(device_id=0)

input_torch = torch.randn(1, 4, 4, 4, 128, dtype=torch.bfloat16)
input_ttnn = ttnn.from_torch(
    input_torch, dtype=ttnn.DataType.BFLOAT16, layout=ttnn.Layout.ROW_MAJOR, device=device
)

padding = [(0, 0), (1, 1), (0, 0), (0, 0), (0, 0)]
output_ttnn = ttnn.pad(input_ttnn, padding, 0.0)

print(output_ttnn)

ttnn.close_device(device)
```

### LINKED ISSUE #53461: pad: rank > 4 shape restore drops padding on upper dimensions
item 2 from master ticket

## Problem

After device padding on a rank > 4 tensor, `update_original_shape()` rebuilds the output logical shape by copying **input** sizes for every dimension above the last two, and only taking the padded sizes from the **last two** dimensions.

Padding applied on dimension `rank-3` (or any upper dim the op allows) is **silently discarded** on restore. The returned tensor can report the wrong shape and volume.

## Example

Rank-5 tensor shape `(2, 3, 4, 5, 6)`, pad dimension `-3` by 1:

| | Shape |
|--|-------|
| Expected padded logical shape | `(2, 3, 5, 5, 6)` — volume 900 |
| Current restore | `(2, 3, 4, 5, 6)` — volume 720 |

The buffer holds 900 elements but the tensor is described as 720 — wrong answer or volume mismatch on downstream ops.

## Location

`ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp` — `update_original_shape()`

## Suggested fix

When rebuilding the logical shape, apply the full padding spec across all dimensions — not just the last two. For dims above the bottom three, derive sizes from the padded shape (or compute logical shape directly from input shape + padding).

## Test plan

Add unit tests in `tests/ttnn/unit_tests/operations/data_movement/test_pad.py` for rank ≥ 5 padding on dimension `rank-3` and other upper dims, comparing shape, volume, and values against PyTorch.

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
Fixes https://github.com/tenstorrent/tt-metal/issues/38144.

`ttnn.pad` rejected padding on any dimension above the lowest three once rank exceeded four, even though the op is documented as rank-generic. The device path already squeezed to 4D; the real gaps were (1) leading dimensions merged by `squeeze_from_ND_to_4D` had no padding mechanism, and (2) ND shape restore dropped padding on upper dims. This PR removes the rank > 4 restriction for device tensors, adds a reshape-based path for leading dims, fixes logical shape restore, and keeps the Quasar mirror in sync.

Host tensors with rank > 4 still only support padding on the lowest three dimensions (`TT_FATAL`), since host tile padding is not fixed here. When dim `rank-4` maps cleanly to 4D axis 0 after squeeze, that dim is handled in the final 4D pass to avoid an extra copy; otherwise it goes through the reshape path. Added rank > 4 unit tests (row-major, tile, dtypes, memory configs, multicore).

Pipeline status

- [x]  Sanity
- [x]  (Runtime) Sanity Test
- [x]  BHPC
- [x]  Nightly for DM Category
- [x]  All model test
- [x]  Single card
- [x]  T3K
- [x]  Galaxy
- [x]  (Tier 1) Models Device Perf Tests
- [x]  (TM-DM) Data Movement Test Suite

 
### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:virdhatchani/38144)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:virdhatchani/38144)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:virdhatchani/38144)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:virdhatchani/38144)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:virdhatchani/38144)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:virdhatchani/38144)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:virdhatchani/38144)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp [hunk 1]
```diff
@@ -85,15 +85,78 @@ ttnn::Tensor pad_via_interleaved_composite(
...
+
+    uint32_t inner_extent = 1;
+    for (int i = dim + 1; i < rank - 2; i++) {
+        inner_extent *= shape[i];
+    }
+
+    ttsl::SmallVector<PadSpecDim> padding_4d = {
+        {0, 0}, {pad_spec.before_elements * inner_extent, pad_spec.after_elements * inner_extent}, {0, 0}, {0, 0}};
+
+    auto padded = ttnn::pad(reshaped, padding_4d, value, use_multicore, memory_config_arg, sub_core_grids);
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp]: `memory_config_arg`/`sub_core_grids` are forwarded to every intermediate pad, and this runs once per padded leading dim (`rank5_all_dims`, `rank6_dim0_dim2_dim5`) — a sharded config is sized for the final rank-N output, so it is wrong for the 4D intermediates, and when this path consumes all the padding the trailing `reshape` returns without re-applying it. Suggest `std::nullopt` for the intermediates and applying the requested config once at the end; `test_pad_rank_gt4_memory_config` only uses interleaved DRAM/L1 and only asserts `buffer_type`, so it won't catch this.

File: tests/ttnn/unit_tests/operations/data_movement/test_pad.py [hunk 2]
```diff
@@ -1284,3 +1277,245 @@ def test_pad_nd_sharded_front_padding_tile_layout_not_supported(device, dtype):
...
+    shape = (1, 4, 4, 32, 128)
+
+    torch_input_tensor = torch.randn(shape, dtype=torch.bfloat16)
+    torch_output_tensor = torch.nn.functional.pad(torch_input_tensor, _ttnn_padding_to_torch(padding), value=0.0)
+
+    input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.bfloat8_b, layout=ttnn.TILE_LAYOUT, device=device)
+    output_tensor = ttnn.to_torch(ttnn.pad(input_tensor, padding, 0.0))
+
+    assert output_tensor.shape == torch_output_tensor.shape
+    assert_with_pcc(torch_output_tensor.float(), output_tensor.float(), 0.999)
```
Comment (Aswinmcw) [inline tests/ttnn/unit_tests/operations/data_movement/test_pad.py]: PCC 0.999 over the whole tensor is a weak assertion for a pure data-movement op, and both parametrized cases fill with 0.0, so a misplaced block would still correlate. Prefer a bit-exact/ULP comparison of the copied region plus an exact check that the padded region equals the fill value — every other new test in this block uses `torch.equal`.

**Review by Aswinmcw — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp [hunk 3]
```diff
@@ -85,15 +85,73 @@ ttnn::Tensor pad_via_interleaved_composite(
...
+
+    uint32_t inner_extent = 1;
+    for (int i = dim + 1; i < rank - 2; i++) {
+        inner_extent *= shape[i];
+    }
+
+    ttsl::SmallVector<PadSpecDim> padding_4d = {
+        {0, 0}, {pad_spec.before_elements * inner_extent, pad_spec.after_elements * inner_extent}, {0, 0}, {0, 0}};
+
+    auto padded = ttnn::pad(reshaped, padding_4d, value, use_multicore, std::nullopt, std::nullopt);
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp]: Dropping `memory_config` here is right, but `sub_core_grids` is a core-allocation constraint rather than a shape-dependent config, so it should still be forwarded — as written, a caller that restricts pad to a sub-grid gets intermediate leading-dim pads running on the full grid. Suggest keeping the `sub_core_grids` parameter on `pad_leading_dimension_via_reshape`/`pad_leading_dimensions` and passing only `std::nullopt` for the memory config.

**Review by Aswinmcw — APPROVED** [review]
Both points addressed — memory config is now applied once on the final tensor (with an `all_leading_dims` case added), and the bfloat8_b test asserts the fill regions exactly plus a bounded body comparison.

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp [hunk 4]
```diff
@@ -85,15 +85,75 @@ ttnn::Tensor pad_via_interleaved_composite(
...
+        dim,
+        rank);
+
+    uint32_t before = 1;
+    for (int i = 0; i < dim; i++) {
+        before *= shape[i];
+    }
+
+    uint32_t middle = 1;
+    for (int i = dim; i < rank - 2; i++) {
```
Comment (amorrisonTT) [inline ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp]: Nit: these loops are a good opportunity to use std::accumulate

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp [hunk 5]
```diff
@@ -268,6 +328,22 @@ ttnn::Tensor pad_impl(
         padding_size = padding.size();
         extra_index = 0;
     }
+    if (original_rank > 4) {
+        // Only padding[extra_index .. extra_index + 3] reaches the 4D kernel. Dims below extra_index have no
+        // slot in the 4D spec at all, and axis 0 of the squeezed tensor may be several original dims folded
+        // together -- padding applied there would stretch the whole merged axis instead of one dimension.
+        // pad_leading_dimensions() consumes those dims before we get here; this guard is what the removed
```
Comment (amorrisonTT) [inline ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp]: Let's drop the Claude-slop historical context from this comment. It can be found via git history if necessary

File: ttnn/cpp/ttnn/operations/experimental/quasar/pad/pad.cpp [hunk 6]
```diff
@@ -28,15 +28,77 @@ using ttnn::operations::data_movement::squeeze_or_unsqueeze_shape_to_ND;
...
+    const int leading_dims_end = rank - 3;
+    for (int dim = 0; dim < leading_dims_end; dim++) {
+        if (padding[dim].before_elements == 0 && padding[dim].after_elements == 0) {
+            continue;
+        }
+        result = pad_leading_dimension_via_reshape(result, dim, padding[dim], value, use_multicore, sub_core_grids);
+        padding[dim] = {0, 0};
+    }
+    return result;
 }
```
Comment (amorrisonTT) [inline ttnn/cpp/ttnn/operations/experimental/quasar/pad/pad.cpp:108]: Same comments for the quasar version

**Review by amorrisonTT — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/32034379617 - passed

**Comment by fplavecTT** [comment]
Please don't leave AI-generated PR description as is, because it is usually overly verbose. For example, there is almost never a need for PR description to list all the changed files, since GH provides that functionality. There is also no need to have two different checklists for pipelines. In general, 2-3 paragraphs is plenty for a PR description.

File: tests/ttnn/unit_tests/operations/data_movement/test_pad.py [hunk 7]
```diff
@@ -613,9 +608,7 @@ def _unsqueeze(smaller, larger, fill):
 )
 @pytest.mark.parametrize("dtype", [ttnn.bfloat16, ttnn.int32, ttnn.uint16, ttnn.bfloat8_b])
 def test_pad_tile(shape, padding, dtype, device):
-    if (shape, padding) in [([5, 4, 3, 2, 1], [1, 0, 0, 0]), ([5, 4, 3, 2, 1], [32, 32, 32, 64])]:
-        pytest.xfail("Can't pad upper dims with rank>4")
-
+    # Padding upper dims of a rank>4 tensor used to xfail; supported since issue #38144.
```
Comment (fplavecTT) [inline tests/ttnn/unit_tests/operations/data_movement/test_pad.py]: These kind of AI-generated historical comments are not useful. Comments should describe current state of code and tests (not trace its history), so they are useful to a future reader.

File: tests/ttnn/unit_tests/operations/data_movement/test_pad.py [hunk 8]
```diff
@@ -1284,3 +1277,280 @@ def test_pad_nd_sharded_front_padding_tile_layout_not_supported(device, dtype):
             use_multicore=True,
             memory_config=output_memory_config,
         )
+
+
+def _ttnn_padding_to_torch(padding):
+    """Convert ttnn per-dim (before, after) padding to torch.nn.functional.pad format."""
+    return tuple(v for i in reversed(range(len(padding))) for v in padding[i])
```
Comment (fplavecTT) [inline tests/ttnn/unit_tests/operations/data_movement/test_pad.py:1341]: This one liner could use a code comment explaining why this non-trivial operation is needed.

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp [hunk 9]
```diff
@@ -454,23 +532,26 @@ ttnn::Tensor pad(
         return input_tensor;
     }
 
+    ttnn::Tensor working_tensor = input_tensor;
     if (original_rank > 4) {
-        const auto first_pad_idx =
-            std::find_if(
-                working_padding.begin(), working_padding.end(), [](auto& p) { return p.after_elements != 0; }) -
-            working_padding.begin();
-        TT_FATAL(
```
Comment (fplavecTT) [inline ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp:462]: Isn't this `TT_FATAL` now removed even for host tensors, which are not fixed (according to PR description)? If so, it would be better to leave `TT_FATAL` active for host tensors.

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp [hunk 10]
```diff
@@ -85,15 +86,69 @@ ttnn::Tensor pad_via_interleaved_composite(
 
 bool eq_spans(const auto a, const auto b) { return std::equal(a.begin(), a.end(), b.begin(), b.end()); }
 
-ttnn::Shape update_original_shape(const ttnn::Shape& padded_shape, const ttnn::Shape& input_shape) {
-    ttsl::SmallVector<uint32_t> updated_shape;
-    size_t input_rank = input_shape.rank();
-    for (size_t i = 0; i < input_rank - 2; i++) {
-        updated_shape.push_back(input_shape[i]);
+ttnn::Shape compute_padded_logical_shape(const ttnn::Shape& input_shape, const ttsl::SmallVector<PadSpecDim>& padding) {
```
Comment (fplavecTT) [inline ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp:87]: All the newly added helpers that are not exposed in the header should go into the anonymous namespace above.

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp [hunk 11]
```diff
@@ -268,6 +323,19 @@ ttnn::Tensor pad_impl(
         padding_size = padding.size();
         extra_index = 0;
     }
+    if (original_rank > 4) {
+        // Only padding[extra_index .. extra_index + 3] reaches the 4D kernel. Dims below extra_index have no
+        // slot in the 4D spec, and axis 0 of the squeezed tensor may fold several original dims together.
+        const auto is_unpadded = [](const PadSpecDim& p) { return p.before_elements == 0 && p.after_elements == 0; };
+        const bool axis0_is_merged =
+            input_tensor_4D.logical_shape()[0] != input_tensor.logical_shape()[static_cast<int>(extra_index)];
+        TT_FATAL(
```
Comment (fplavecTT) [inline ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp:350]: The leading-dimension path does one full device pad per dimension, so padding two leading dims plus a trailing dim costs three complete copies of the data. However, the 4D kernel accepts padding amounts for all four slots in a single call, but each round here fills in only slot 1 and passes {0,0} for the rest, and the final trailing-dims round fills slots 1 - 3 while leaving slot 0 at zero. So there's a free slot in the last round.
Could we take advantage of this to make it more efficient and reduce number of copies?

Same would apply to Quasar version.

File: tests/ttnn/unit_tests/operations/data_movement/test_pad.py (hunk 7 above)
Comment (VirdhatchaniKN) [reply tests/ttnn/unit_tests/operations/data_movement/test_pad.py]: Removed historical/issue-reference comments from the rank > 4 test section and trimmed verbose docstrings.

File: tests/ttnn/unit_tests/operations/data_movement/test_pad.py (hunk 8 above)
Comment (VirdhatchaniKN) [reply tests/ttnn/unit_tests/operations/data_movement/test_pad.py:1341]: Added a comment explaining that torch.nn.functional.pad expects padding pairs in reverse dimension order (last dim first).

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp (hunk 9 above)
Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp:462]: Restored TT_FATAL for host tensors with rank > 4; device tensors use the leading-dim reshape path.

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp (hunk 10 above)
Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp:87]: Moved eq_spans, compute_padded_logical_shape, pad_leading_dimension_via_reshape, and pad_leading_dimensions into an anonymous namespace. Exposed apply_leading_dimension_padding as the single detail entry point.

File: ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp (hunk 11 above)
Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/data_movement/pad/pad.cpp:350]: Dim rank-4 is deferred to the final 4D pass only when squeezed axis 0 equals shape[rank-4] (not merged with earlier dims). Otherwise it still uses the reshape path. Mirrored in Quasar.

**Comment by VirdhatchaniKN** [comment]
Addressed the latest review feedback in 26ee65fb9a8 and shortened the PR description.

**Review by fplavecTT — APPROVED** [review]

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/32350462318 - passed
